### PR TITLE
#168781709 - like/unlike accommodation facility 

### PR DESCRIPTION
**Pull request by ngireric123**
#### What does this PR do?

 like/unlike an accommodation facility

#### Description of Task to be completed?
- Allow users to like/unlike accommodation facilities.

##### like/unlike accommodation
1. Clone this repository
2. Run ` sequelize db:migrate `
3. Run ` sequelize db:seed:All `
4. Run ` npm run dev `
5. Log into the application at `api/v1/users/login`
6. Using an  authenticated user send a POST request to `/api/v1/accommodations/${id}/like`
7. It should like/unlike accommodation facilities. 

##### Get number of likes

1. Using an  authenticated user send a POST request to `/api/v1/accommodations/${slug}`
2. It should display details of accommodation and its likes

#### Any background context you want to provide?

N / A

#### What are the relevant pivotal tracker stories?
[#168781709](https://www.pivotaltracker.com/story/show/168781709)
#### Screenshots (if appropriate)

image

image

#### Questions:

N/A

### CODE REVIEW

**Review by Cheza-Dzabala — CHANGES_REQUESTED** [review]
Hey Eric. Well done. However, please use lowercase for the DB table names not uppercase. Secondly, the error message returned should be refactored to something like: "The specified accommodation does not exist in our system." Otherwise, great work.

**Review by sabin18 — APPROVED** [review]
Good job @ngireric123 .

File: src/routes/api/accommodations.js [hunk 1]
```diff
@@ -205,6 +202,6 @@ router.patch('/book', validateToken, checkBookingRole, validateBooking, bookAcco
 router.get('/:slug', validateToken, viewSpecificAccommodation);
 router.patch('/activate/:slug', validateToken, validateReasons, accommodationActivation);
 router.get('/admin/deactivated', validateToken, viewDeactivated);
-
+router.post('/:id/like', validateToken, checkId, likeValidation, likeAccommodation);
```
Comment (nakiwuge) [inline src/routes/api/accommodations.js]: Make this route dynamic to cater for like and dislike you can add something like this `/:id/:action`

File: src/services/accommodationServices.js [hunk 2]
```diff
@@ -6,4 +6,13 @@ export default class AccommodationServices {
       where: property
     }).then(accommodation => accommodation);
   }
+
+  static async findIdAccommodation(id) {
+    const accommodation = await models.accommodations.findOne({
+      where: { id }
+    });
+    if (!accommodation) return null;
```
Comment (nakiwuge) [inline src/services/accommodationServices.js]: Move this condition inside the then on line 7 and do away with this function

File: src/services/likesServices/likesServices.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import models from '../../database/models';
+
+class LikeService {
+  static async findLikes(accommodation, userId) {
+    const like = await models.Likes.findOne({
+      where: { accommodationId: accommodation, user_id: userId }
```
Comment (nakiwuge) [inline src/services/likesServices/likesServices.js]: This condition will only allow users who have liked and accommodation to view the total number of likes in an accommodation. We should allow users to view the total number of likes/dislikes on an accommodation whether they have liked that accommodation or not

File: src/services/likesServices/likesServices.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return like.dataValues;
+  }
+
+  static async addNewLike(like) {
+    const newLike = await models.Likes.create(like);
+    return newLike;
+  }
+
+  static async updateLike(accommodation, userId, status) {
+    await models.Likes.update(status, {
```
Comment (nakiwuge) [inline src/services/likesServices/likesServices.js]: You should only update when the passed param is `dislike`

File: src/services/likesServices/likesServices.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      where:
+            { accommodationId: accommodation, user_id: userId }
+    });
+    return status;
+  }
+
+  static async likesSum(slug) {
+    const idAccommodation = await models.accommodations.findOne({
```
Comment (nakiwuge) [inline src/services/likesServices/likesServices.js]: rename `idAccommodation` to `accomodation`

File: src/services/likesServices/likesServices.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      where: { slug }
+    });
+    if (idAccommodation) {
+      const accommodation = idAccommodation.id;
```
Comment (nakiwuge) [inline src/services/likesServices/likesServices.js]: rename `accommodation` to  `accommodationId`

File: src/tests/accommodationTest.spec.js [hunk 7]
```diff
@@ -161,6 +162,9 @@ describe('Accommodation Test', () => {
       .set('Authorization', `Bearer ${requesterToken}`)
       .end((err, res) => {
         res.should.have.status(200);
+      console.log(err);
```
Comment (nakiwuge) [inline src/tests/accommodationTest.spec.js]: remove the console

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@ngireric123 great job on the implementation kindly works on the feedback below.
- Add total number of dislikes when fetching accomodations and a single accommodation.
- Add `hasLiked` and `hasDisliked` boolean keys when fetching accommodations  and a single accommodation 
Also check on the feedback I have left on this pr

File: src/database/migrations/20191105100202-create-likes.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Likes', {
```
Comment (Cheza-Dzabala) [inline src/database/migrations/20191105100202-create-likes.js]: Make tables lowercase

File: src/database/models/accommodations.js [hunk 9]
```diff
@@ -1,39 +1,48 @@
...
+    { tableName: "accommodations" }
+  );
   accommodations.associate = models => {
     // associations can be defined here
     accommodations.belongsTo(models.users, {
-      as: 'ownerUser',
-      foreignKey: 'owner',
-      targetKey: 'id',
+      as: "ownerUser",
+      foreignKey: "owner",
```
Comment (Cheza-Dzabala) [inline src/database/models/accommodations.js:20]: This should be ` userId ` not ` owner `

File: src/database/models/accommodations.js [hunk 10]
```diff
@@ -1,39 +1,48 @@
...
+    {
+      name: DataTypes.STRING,
+      description: DataTypes.STRING,
+      locationId: DataTypes.INTEGER,
+      availableSpace: DataTypes.INTEGER,
+      cost: DataTypes.INTEGER,
+      currency: DataTypes.STRING,
+      highlights: DataTypes.STRING,
+      amenities: DataTypes.STRING,
+      owner: DataTypes.INTEGER,
```
Comment (Cheza-Dzabala) [inline src/database/models/accommodations.js]: This should be userId not owner

File: src/database/models/accommodations.js [hunk 11]
```diff
@@ -1,39 +1,48 @@
...
+    },
+    { tableName: "accommodations" }
+  );
   accommodations.associate = models => {
     // associations can be defined here
     accommodations.belongsTo(models.users, {
-      as: 'ownerUser',
-      foreignKey: 'owner',
-      targetKey: 'id',
+      as: "ownerUser",
```
Comment (Cheza-Dzabala) [inline src/database/models/accommodations.js:19]: This should just be `owner`. Not `ownerUser`

File: src/database/models/accommodations.js [hunk 12]
```diff
@@ -1,39 +1,48 @@
...
+      foreignKey: "owner",
+      targetKey: "id"
     });
     accommodations.belongsTo(models.locations, {
-      as: 'accommodationLocation',
-      foreignKey: 'locationId',
-      targetKey: 'id',
+      as: "accommodationLocation",
```
Comment (Cheza-Dzabala) [inline src/database/models/accommodations.js:24]: Relation name should be ` location `. Not  ` accommodationLocation `

File: src/database/models/likes.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const Likes = sequelize.define('Likes', {
```
Comment (Cheza-Dzabala) [inline src/database/models/likes.js]: Lowercase table names only

**Review by Cheza-Dzabala — CHANGES_REQUESTED** [review]
Please see comments on PR.

**Review by nignanthomas — APPROVED** [review]
@ngireric123 
Great job man
Rien à signaler

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**